### PR TITLE
Rancher Monitoring Support

### PR DESCRIPTION
**Pull request by orangedeng**
Related Issue: https://github.com/rancher/rancher/issues/14230
Related types PR: https://github.com/rancher/types/pull/636

This is a PR for rancher monitoring support. Summaries are follow:

1. Adding new kind of Alert feature that supports grouping alert rules.
2. Support prometheus crds in Rancher API.
3. Enable and disable cluster and project monitoring.
4. Support monitoring metrics graph.
5. Support prometheus to collect workload metrics with operator.

### CODE REVIEW

**Comment by cjellick** [comment]
Build is not greeen. Looks related

**Comment by cjellick** [comment]
It looks like we have put the Prometheus namespace into the project, rather than into the system project, which is what we discussed/requested. Is that correct? What is the technical need for that? How are you handling RBAC? As I understand it, Prometheus requires escalated privileges to the host so we dont want project owners or project members to have access to it. Have you achieved that?

File: pkg/api/server/managementstored/setup.go [hunk 1]
```diff
@@ -103,6 +110,14 @@ func Setup(ctx context.Context, apiContext *config.ScaledContext, clusterManager
 		projectclient.SourceCodeRepositoryType,
 	)
 
+	// create Prometheus Operator CRD
```
Comment (cjellick) [inline pkg/api/server/managementstored/setup.go]: you create these in two places? Here and in `clusterdeploiy.go`. Why?

File: scripts/test [hunk 2]
```diff
@@ -12,6 +12,8 @@ cd $(dirname $0)/..
 
 CMD=bin/rancher
 
+ln -s "$PWD/bin/rancher" /usr/bin/kubectl
```
Comment (cjellick) [inline scripts/test]: wy do you need to do this?

File: pkg/systemtemplate/template.go [hunk 3]
```diff
@@ -197,4 +197,109 @@ spec:
           secretName: cattle-credentials-{{.TokenKey}}
   updateStrategy:
     type: RollingUpdate
+---
+apiVersion: apiextensions.k8s.io/v1beta1
```
Comment (cjellick) [inline pkg/systemtemplate/template.go]: How are these CRDs related or not realted to the ones you create via BatchCreateCRDs in clusterdeploy.go

**Comment by cjellick** [comment]
> Adding new kind of Alert feature that supports grouping alert rules. 

Is this directly and fundamentally tied to the monitoring feature? If not, it should not have been included in this PR. The PR is huge. Combining multiple features into a single PR makes it much more difficult to review. I see that the types PR also combines these features, so it is probably too difficult to disentangle into separate PRs now, but in the future, please don't combine multiple features into a single PR.

**Comment by cjellick** [comment]
Tried to run this PR locally and got this error:
```
Failed to ensure monitoring project name for Cluster local(local): failed to move Namespace cattle-prometheus to Project System: Operation cannot be fulfilled on namespaces "cattle-prometheus": the object has been modified; please apply your changes to the latest version and try again
```
prom-error

File: pkg/api/customization/project/project_actions.go [hunk 4]
```diff
@@ -26,6 +28,8 @@ import (
 func Formatter(apiContext *types.APIContext, resource *types.RawResource) {
 	resource.AddAction(apiContext, "setpodsecuritypolicytemplate")
 	resource.AddAction(apiContext, "exportYaml")
+	resource.AddAction(apiContext, "enableMonitoring")
```
Comment (cjellick) [inline pkg/api/customization/project/project_actions.go]: logic should be smarter such that `enableMonitoring` is only showed when montioring is currently disabled and `disbleMonitoring` is only shown when its currently enabled. The UI will decide to hide or show these options based off of the presence or absence of these actions

**Comment by cjellick** [comment]
From our conversation with Darren, change the way you create CRDs. Do not create them in the system template or clusterdeploy.go

Follow the paradigm darren uses in rio:
    ```version = types.APIVersion{
        Version: "v1alpha3",
        Group:   "networking.istio.io",
        Path:    "/v1alpha3-istio-networking",
    }

    schemas = factory.Schemas(&version).
        MustImport(&version, v1alpha3.Gateway{}).
        MustImport(&version, v1alpha3.VirtualService{}).
        MustImport(&version, v1alpha3.DestinationRule{})
    //server := norman.GetServer(ctx)
    factory := crd.NewFactoryFromClientGetter(clientGetter)
    // We create istio types so that our controllers don't error on first start
    _, err := factory.CreateCRDs(ctx, normantypes.DefaultStorageContext,
        schemas.Schema(&version, v1alpha3.GatewayGroupVersionKind.Kind),
        schemas.Schema(&version, v1alpha3.VirtualServiceGroupVersionKind.Kind),
        schemas.Schema(&version, v1alpha3.DestinationRuleGroupVersionKind.Kind))
    if err != nil {
        return err
    }

    factory.BatchWait()```
the catch is I just create a temporary APIVersion and Schema for the purpose of creating the CRDs
This logic should live in this file: `pkg/controllers/user/controller.go`

**Comment by cjellick** [comment]
LGTM

**Comment by loganhz** [comment]
Link to https://github.com/rancher/rancher/issues/15706
